libkbfs/quota_usage.go

```diff
@@ -0,0 +1,134 @@
+// Copyright 2017 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package libkbfs
+
+import (
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"golang.org/x/net/context"
+
+	"github.com/keybase/client/go/logger"
+	"github.com/keybase/kbfs/kbfsblock"
+)
+
+//ECQUCtxTagKey is the type for unique ECQU background opertaion IDs.
+type ECQUCtxTagKey struct{}
+
+// ECQUID == "ECQU"
+const ECQUID = "ECQU"
+
+type cachedQuotaUsage struct {
+	timestamp  time.Time
+	usageBytes int64
+	limitBytes int64
+}
+
+// EventuallyConsistentQuotaUsage keeps tracks of quota usage, in a way user of
+// which can choose to accept stale data to reduce calls into block servers.
+type EventuallyConsistentQuotaUsage struct {
+	config Config
+	log    logger.Logger
+
+	backgroundInProcess int32
+
+	mu     sync.RWMutex
+	cached cachedQuotaUsage
+}
+
+// NewEventuallyConsistentQuotaUsage creates a new
+// EventuallyConsistentQuotaUsage object.
+func NewEventuallyConsistentQuotaUsage(
+	config Config) *EventuallyConsistentQuotaUsage {
+	return &EventuallyConsistentQuotaUsage{
+		config: config,
+		log:    config.MakeLogger(ECQUID),
+	}
+}
+
+func (q *EventuallyConsistentQuotaUsage) getAndCache(
+	ctx context.Context) (usage cachedQuotaUsage, err error) {
+	defer func() {
+		q.log.CDebugf(ctx, "getAndCache: error=%v", err)
+	}()
+	quotaInfo, err := q.config.BlockServer().GetUserQuotaInfo(ctx)
+	if err != nil {
+		return usage, err
+	}
+
+	usage.limitBytes = quotaInfo.Limit
+	if quotaInfo.Total != nil {
+		usage.usageBytes = quotaInfo.Total.Bytes[kbfsblock.UsageWrite]
+	} else {
+		usage.usageBytes = 0
+	}
+	usage.timestamp = time.Now()
+
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	q.cached = usage
+
+	return usage, nil
+}
+
+// Get returns KBFS bytes used and limit for user. To help avoid having too
+// frequent calls into bserver, caller can provide a positive tolerance, to
+// accept stale LimitBytes and UsageByts data.
+//
+// 1) If the cached (stale) data is older for less than half of tolerance, the
```

Comment:
Grammar nit: "If the age of the cached data is less than half..."  Please fix the other ones below in a similar way.